distributed/scheduler.py

```diff
@@ -8558,39 +8652,132 @@ def check_idle(self) -> float | None:
         return self.idle_since
 
     def _check_no_workers(self) -> None:
-        """Shut down the scheduler if there have been tasks ready to run which have
-        nowhere to run for `distributed.scheduler.no-workers-timeout`, and there
-        aren't other tasks running.
-        """
-        if self.status in (Status.closing, Status.closed):
-            return  # pragma: nocover
-
         if (
-            (not self.queued and not self.unrunnable)
-            or (self.queued and self.workers)
-            or any(ws.processing for ws in self.workers.values())
+            self.status in (Status.closing, Status.closed)
+            or self.no_workers_timeout is None
         ):
-            self._no_workers_since = None
             return
 
-        # 1. There are queued or unrunnable tasks and no workers at all
-        # 2. There are unrunnable tasks and no workers satisfy their restrictions
-        # (Only rootish tasks can be queued, and rootish tasks can't have restrictions)
+        now = monotonic()
+        stimulus_id = f"check-no-workers-timeout-{time()}"
 
-        if not self._no_workers_since:
-            self._no_workers_since = time()
-            return
+        recommendations: Recs = {}
 
-        if (
-            self.no_workers_timeout
-            and time() > self._no_workers_since + self.no_workers_timeout
-        ):
-            logger.info(
-                "Tasks have been without any workers to run them for %s; "
-                "shutting scheduler down",
-                format_time(self.no_workers_timeout),
+        self._refresh_no_workers_since(now)
+
+        affected = self._check_unrunnable_task_timeouts(
+            now, recommendations=recommendations, stimulus_id=stimulus_id
+        )
+
+        affected.update(
+            self._check_queued_task_timeouts(
+                now, recommendations=recommendations, stimulus_id=stimulus_id
             )
-            self._ongoing_background_tasks.call_soon(self.close)
+        )
+        self.transitions(recommendations, stimulus_id=stimulus_id)
+        if affected:
+            self.log_event(
+                "scheduler",
+                {"action": "no-workers-timeout-exceeded", "keys": affected},
+            )
+
+    def _check_unrunnable_task_timeouts(
+        self, timestamp: float, recommendations: Recs, stimulus_id: str
+    ) -> set[Key]:
+        assert self.no_workers_timeout
+        unsatisfied = []
+        no_workers = []
+        for ts, unrunnable_since in self.unrunnable.items():
+            if timestamp <= unrunnable_since + self.no_workers_timeout:
+                break
```

Comment:
shouldn't this be a continue?

Reply:
Notes from an offline conversation:

`unrunnable` is a dict and thus insertion-ordered, which guarantees that `unrunnable_since` will be monotonically increasing. I'll add a comment and an optional validation step to ensure that `unrunnable` _remains_ insertion-ordered.